src/lib/x509/x509_ext.cpp

```diff
@@ -868,6 +877,582 @@ const std::string& TNAuthList::Entry::telephone_number() const {
+   }
+
+   into.end_cons();
+}
+
+void ASBlocks::ASIdentifiers::decode_from(Botan::BER_Decoder& from) {
+   BER_Object obj = from.get_next_object();
+   ASN1_Type type_tag = obj.type_tag();
+   if(type_tag != ASN1_Type::Sequence) {
+      throw Decoding_Error(fmt("Unexpected type for ASIdentifiers {}", static_cast<uint32_t>(type_tag)));
+   }
+   BER_Decoder obj_ber = BER_Decoder(obj);
+
+   BER_Object elem_obj = obj_ber.get_next_object();
+   uint32_t elem_type_tag = static_cast<uint32_t>(elem_obj.type_tag());
+
+   // asnum, potentially followed by an rdi
+   if(elem_type_tag == 0) {
+      BER_Decoder as_obj_ber = BER_Decoder(elem_obj);
+      ASIdentifierChoice asnum;
+      as_obj_ber.decode(asnum);
+      m_asnum = asnum;
+
+      BER_Object rdi_obj = obj_ber.get_next_object();
+      ASN1_Type rdi_type_tag = rdi_obj.type_tag();
+      if(static_cast<uint32_t>(rdi_type_tag) == 1) {
+         BER_Decoder rdi_obj_ber = BER_Decoder(rdi_obj);
+         ASIdentifierChoice rdi;
+         rdi_obj_ber.decode(rdi);
+         m_rdi = rdi;
+      } else if(rdi_type_tag != ASN1_Type::NoObject) {
+         throw Decoding_Error(fmt("Unexpected type for ASIdentifiers rdi: {}", static_cast<uint32_t>(rdi_type_tag)));
+      }
+   }
+
+   // just an rdi
+   if(elem_type_tag == 1) {
+      BER_Decoder rdi_obj_ber = BER_Decoder(elem_obj);
+      ASIdentifierChoice rdi;
+      rdi_obj_ber.decode(rdi);
+      m_rdi = rdi;
+      BER_Object end = obj_ber.get_next_object();
+      ASN1_Type end_type_tag = end.type_tag();
+      if(end_type_tag != ASN1_Type::NoObject) {
+         throw Decoding_Error(
+            fmt("Unexpected element with type {} in ASIdentifiers", static_cast<uint32_t>(end_type_tag)));
+      }
+   }
+}
+
+void ASBlocks::ASIdentifierChoice::encode_into(Botan::DER_Encoder& into) const {
+   if(m_as_ranges.has_value()) {
+      into.start_sequence().encode_list(m_as_ranges.value()).end_cons();
+   } else {
+      into.encode_null();
+   }
+}
+
+void ASBlocks::ASIdentifierChoice::decode_from(Botan::BER_Decoder& from) {
+   BER_Object obj = from.get_next_object();
+   ASN1_Type type_tag = obj.type_tag();
+
+   if(type_tag == ASN1_Type::Null) {
+      m_as_ranges = std::nullopt;
+   } else if(type_tag == ASN1_Type::Sequence) {
+      BER_Decoder obj_ber = BER_Decoder(obj);
+      std::vector<ASIdOrRange> as_ranges;
+
+      while(obj_ber.more_items()) {
+         ASIdOrRange as_id_or_range;
+         obj_ber.decode(as_id_or_range);
+         as_ranges.push_back(as_id_or_range);
+      }
+
+      m_as_ranges = as_ranges;
+      sort_and_merge();
+   } else {
+      throw Decoding_Error(fmt("Unexpected type for ASIdentifierChoice {}", static_cast<uint32_t>(type_tag)));
+   }
+}
+
+void ASBlocks::ASIdentifierChoice::sort_and_merge() {
+   if(!m_as_ranges.has_value()) {
+      return;
+   }
+   std::vector<ASIdOrRange> current = m_as_ranges.value();
+   if(current.empty()) {
+      return;
+   }
+   bool changed = true;
+   while(changed) {
```

Comment:
Can this and the other `sort_and_merge` logic be pulled out to a new function (in an anonymous namespace in this file) that handles both? They seem nearly identical